File: backend/app/api/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.db.database import SessionLocal
from app.models.notification import Notification
from app.models.users import User
from app.api.auth import get_current_user
from app.schemas.notification import NotificationResponse


router = APIRouter(prefix="/api/notifications", tags=["notifications"])


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("", response_model=list[NotificationResponse])
def list_notifications(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List notifications for the current user (latest first)."""
    items = (
        db.query(Notification)
        .filter(Notification.user_id == current_user.id)
        .order_by(Notification.created_at.desc(), Notification.id.desc())
        .all()
    )
    return items


@router.put("/{notification_id}/read", response_model=NotificationResponse)
def mark_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    item = (
        db.query(Notification)
        .filter(Notification.id == notification_id, Notification.user_id == current_user.id)
        .first()
    )
    if not item:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    item.is_read = True
    db.commit()
    db.refresh(item)
    return item


@router.put("/read-all", status_code=status.HTTP_204_NO_CONTENT)
def mark_all_read(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    db.query(Notification).filter(
        Notification.user_id == current_user.id, Notification.is_read == False  # noqa: E712
    ).update({"is_read": True})
    db.commit()
    return None


@router.delete("/{notification_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_notification(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    item = (
        db.query(Notification)
        .filter(Notification.id == notification_id, Notification.user_id == current_user.id)
        .first()
    )
    if not item:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    db.delete(item)
    db.commit()
    return None

File: frontend1/src/services/notifications.ts
import api from "./api";

export type NotificationType =
  | "HIGH_RISK"
  | "FOLLOW_UP"
  | "REPORT_READY"
  | "NEW_PATIENT_ASSIGNED"
  | string;

export type NotificationResponse = {
  id: number;
  user_id: number;
  patient_id: number | null;
  report_id: number | null;
  type: NotificationType;
  title: string;
  message: string;
  is_read: boolean;
  created_at: string;
};

export async function listNotifications() {
  const { data } = await api.get<NotificationResponse[]>("/notifications");
  return data;
}

export async function markNotificationRead(id: number) {
  const { data } = await api.put<NotificationResponse>(`/notifications/${id}/read`);
  return data;
}

export async function markAllNotificationsRead() {
  await api.put("/notifications/read-all");
}

export async function deleteNotification(id: number) {
  await api.delete(`/notifications/${id}`);
}

